src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborWriter.Map.cs

```diff
@@ -3,49 +3,233 @@
+using System.Diagnostics;
+using System.Threading;
 
 namespace System.Security.Cryptography.Encoding.Tests.Cbor
 {
     internal partial class CborWriter
     {
+        private static readonly ArrayPool<KeyValueEncodingRange> s_rangeBuffers = ArrayPool<KeyValueEncodingRange>.Create();
+
+        private KeyEncodingComparer? _keyComparer;
+        private Stack<HashSet<KeyValueEncodingRange>>? _pooledKeyValueEncodingRangeSets;
+
         public void WriteStartMap(int definiteLength)
         {
-            if (definiteLength < 0)
+            if (definiteLength < 0 || definiteLength > int.MaxValue / 2)
             {
-                throw new ArgumentOutOfRangeException(nameof(definiteLength), "must be non-negative integer.");
+                throw new ArgumentOutOfRangeException(nameof(definiteLength));
             }
 
             WriteUnsignedInteger(CborMajorType.Map, (ulong)definiteLength);
-            AdvanceDataItemCounters();
-            PushDataItem(CborMajorType.Map, 2 * (uint)definiteLength);
+            PushDataItem(CborMajorType.Map, definiteLength: checked(2 * definiteLength));
+        }
+
+        public void WriteStartMap()
+        {
+            EnsureWriteCapacity(1);
+            WriteInitialByte(new CborInitialByte(CborMajorType.Map, CborAdditionalInfo.IndefiniteLength));
+            PushDataItem(CborMajorType.Map, definiteLength: null);
+            _currentKeyOffset = _offset;
+            _currentValueOffset = null;
         }
 
         public void WriteEndMap()
         {
-            if (!_isEvenNumberOfDataItemsWritten)
+            if (_itemsWritten % 2 == 1)
             {
                 throw new InvalidOperationException("CBOR Map types require an even number of key/value combinations");
             }
 
-            bool isDefiniteLengthMap = _remainingDataItems.HasValue;
-
             PopDataItem(CborMajorType.Map);
+            AdvanceDataItemCounters();
+        }
 
-            if (!isDefiniteLengthMap)
+        //
+        // Map encoding conformance
+        //
+
+        private void HandleKeyWritten()
+        {
+            Debug.Assert(_currentKeyOffset != null && _currentValueOffset == null);
+
+            int currentValueOffset = _offset;
+
+            if (CborConformanceLevelHelpers.RequiresUniqueKeys(ConformanceLevel))
             {
-                // append break byte
-                EnsureWriteCapacity(1);
-                _buffer[_offset++] = CborInitialByte.IndefiniteLengthBreakByte;
+                HashSet<KeyValueEncodingRange> ranges = GetKeyValueEncodingRanges();
+
+                var currentKeyRange = new KeyValueEncodingRange(
+                    offset: _currentKeyOffset.Value,
+                    keyLength: currentValueOffset - _currentKeyOffset.Value,
+                    totalLength: 0 // totalLength not known, but is not significant w.r.t. key equality semantics
+                );
+
+                if (ranges.Contains(currentKeyRange))
```

Comment:
I'm guessing that the hash set does value comparison based on stored ranges.  The English here reads as "if we've already stored this range", vs "if we've already seen this key".  I don't know if that means "add a comment" or "see if the variables can be renamed to greater effect", e.g.

```suggestion
                if (keys.Contains(newKey))
```